Add initial (very limited) support for line reduce scatter

SeanNijjar opened the pull request:
# Ticket
#11178 

# Problem description
We don't support line reduce scatter

# What's changed

This commit adds the initial support for reduce scatter. However, only a
few cases are functional. Future work will improve correctness across
more cases.

## Line Reduce Scatter Algorithm
The algorithm for line reduce scatter will send the minimal ammount of
data over each line and out of each chip. All diagrams are for an
example 4-chip line reduce scatter.

First the operation fractures each input tensor

```
Input Tensors ---------------
       |      |      |      |
       |      |      |      |
       v      v      v      v
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|

Chip   0      1      2      3

                 |
                 | Fracture
                 | Tensors
                 v

Input Tensors ---------------
       |      |      |      |
       |      |      |      |
       v      v      v      v
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|

Chip   0      1      2      3
```

With fracture tensors are reduced and collapsed to the diagonal across
the chips where the diagonal shows how the fractures spatially map to
the final outputs. For example, the first output is generated by
reducing the top chunk of each input tensor.

The reduction is performed by having each chip forward its input to its
neighbour. For chips that are not at the end of the line, they reduce
with their input and forward.

```
      |-|    |-|    |-|    |-|
      |#|<---| |<---| |<---| |
      |-|    |-|    |-|    |-|
      | |--->|#|<---| |<---| |
      |-|    |-|    |-|    |-|
      | |--->| |--->|#|<---| |
      |-|    |-|    |-|    |-|
      | |--->| |--->| |--->|#|
      |-|    |-|    |-|    |-|

Chip   0      1      2      3
```

However, note that each arrow from the diagram heading out of of a chip
in a given direction shares ethernet resources for all other arrows
heading in the same direction from that chip. This means there is
inherently serialization here. For that reason, we schedule the chunks
in some way.

The general scheduling strategy is to send the chunks that are furthest
from the final reduce output first and step through chunks that are
incrementally closer to the final output.

Each direction from a chip can be processed independently.

The diagram below is annotated with the "timesteps" when each chunk is
sent. Each timestep is marked relative to the chunk source.

```
      |-| t=0|-| t=0|-| t=0|-|
      |#|<---| |<---| |<---| |
      |-|t=2 |-| t=1|-| t=1|-|
      | |--->|#|<---| |<---| |
      |-|t=1 |-|t=1 |-| t=2|-|
      | |--->| |--->|#|<---| |
      |-|t=0 |-|t=0 |-|t=0 |-|
      | |--->| |--->| |--->|#|
      |-|    |-|    |-|    |-|

Chip   0      1      2      3
```

Finally, not that the final output requires a reduction from both
directions. Given that the two directions of the line are executing
completely indepdently, we require some sort of merge operation. At the
time of this commit, the merge strategy is to designate a master and
slave reducer direction. We arbitrarily choose the 'right' or
'clockwise' direction as the master.

The master direction will write its output to the output tensor but note
that this will only be a partial output. The slave direction will read
from the output tensor to merge with the data from producer chip. It
will read from the output tensor based on a credit passing from master
 (implemented via semaphores)

```
     -------Input Tensor
     |
     |
     |
     |
     |
     |      |---------|----------
     |----> |         |         |
            | Reader  | Sender  |---
  From EDM  | (master)| (master)|  |
 ---------> |         |         |  |
            |---------|---------|  |
                                   |
                                   |
        |------  Output Tensor <---|
        |                 ^
        |                 |---------
        |                          |
        |   |---------|----------  |
        --> |         |         |  |
            | Reader  | Sender  |--|
  From EDM  | (master)| (master)|
 ---------> |         |         |
            |---------|---------|
```

As a part of the line reduce scatter implementation, new CCL
componenst were added: ccl send and ccl command generators/readers.

The ccl_send (kernel) was used to implement the starting ends of the lines
(i.e. the first senders). Although the ccl_send provides more generic
send capabilities than line reduce scatter currently requires, it was
chosen because it is a basic building block also for future CCL
send/recv "operations" and higher level CCL programming models.

## CCL Send (Kernel)
The ccl_send kernel acts like an interpreter of CCL commands. CCL
commands are, so far, limited to be a send from tensor to EDM of a
tensor slice. The command specifies some information about the tensor
(shape, slice/view shape, view offset, etc.).

CCL send is capable of executing multiple commands back to back. In the
context of line reduce scatter, the ccl_send implements the separate
sends of the fractured chunks on the left and right ends of the line. To
do this for a line reduce scatter, we invoke n commands where n=#chips
in the line. Future commands will let an invoker specify this basic
pattern as a single command.

Looking at the third diagram that outlines the timesteps for each chunk,
for the left/right tensors, each timestep directly maps to a separate
ccl command.

## CCL Command Generators/Readers 
To facilitate command generation, initial components have been added to
let the host serialize commands for the the ccl_send kernel.
Correspondingly, command unpacking logic is also specified for each
command. This is used to help simplify command generation for the host.

Note that ccl_send as a standalone kernel and operation is experimental
and has several limitations:
- Slice reads currently constrained to page aligned slices
- Host command generation doesn't support proper 4D shape support
    (although the kernel side will internally represent shapes as 4D)
- Only one command is currently supported (send tensor slice to EDM)

# Future Work:
These are specific to line reduce scatter
1. Resolve correctness bugs (PCC) exposed by non-enabled test cases from ring and enable those tests in the test list
2. Enable sharding support
3. Drive towards minimum op completion requirements which will be the same effort as ring reduce scatter as the pieces that need changing there are common amongst the two (namely proper 4D shape representation on device side - today we only pass a flattened-to-2d shape)


Merging now to
1. Unblock (composite) implementation for all-reduce op 
2. Enable more incremental development of ccl-send-recv op
4. Avoid repeated conflicts due to rebase
 

# Checklist
- [x] Post commit CI: https://github.com/tenstorrent/tt-metal/actions/runs/11043482824
  - https://github.com/tenstorrent/tt-metal/actions/runs/11086517124
- [x] t3000 frequent: https://github.com/tenstorrent/tt-metal/actions/runs/11043294192
- [x] t3000 model perf: https://github.com/tenstorrent/tt-metal/actions/runs/11043404805
  - Not completely green but neither is main. Seeing the same failures as on main and failures are unrelated to changes here. 
- [x] t3000 unit test: https://github.com/tenstorrent/tt-metal/actions/runs/11043407305
- [x] TG frequent: https://github.com/tenstorrent/tt-metal/actions/runs/11043423582
- [x] Blackhole Post commit (if applicable)
- [x] New/Existing tests provide coverage for changes

TT-BrianLiu reviewed (approved):
2024 for new files

cfjchu commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter.cpp:
```diff
@@ -14,11 +14,12 @@ ttnn::Tensor ExecuteReduceScatter::invoke(
     ttnn::operations::reduction::ReduceType math_op,
     const uint32_t num_links,
     const std::optional<ttnn::MemoryConfig>& memory_config,
+    ttnn::ccl::Topology topology,
```
when `ttnn::ccl::Topology::Line` is used, don't we need something similar to `cluster_axis`define how we want to kick off parallel line gathers on MeshTensor?

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/ccl_common.cpp:
```diff
@@ -99,111 +107,8 @@ std::unique_ptr<CclOpTensorConfig> CclOpTensorConfig::build_all_gather_tensor_co
     }
 }
 
-static std::pair<tt_xy_pair, tt_xy_pair> shard_grid_from_shard_spec(const ShardSpec& shard_spec) {
```
Moved to ccl args types emitters

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:
```diff
@@ -40,302 +42,21 @@ namespace ttnn {
-        return args;
-    }
-
-    std::vector<uint32_t> generate_receiver_kernel_rt_args(
```
kernel rt & ct args emitting moved to worker builder cpp

SeanNijjar reviewed (commented):
pointing out moved code

cfjchu commented:
@SeanNijjar and i synced offline, there's a part 2 coming.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_pybind.hpp:
```diff
@@ -5,7 +5,11 @@
 #pragma once
 
 #include "pybind11/pybind_fwd.hpp"
```
You don’t need this include, right?

ayerofieiev-tt replied:
(I mean newly added includes)

SeanNijjar replied:
No - cleaned up

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_pybind.cpp:
```diff
@@ -63,6 +66,7 @@ void py_bind_reduce_scatter(pybind11::module& module) {
         Keyword Args:
             * :attr:`num_links` (int): Number of links to use for the all-gather operation.
             * :attr:`memory_config` (Optional[ttnn.MemoryConfig]): Memory configuration for the operation.
+            * :attr:`topology` (Optional[ttnn.ccl.topology]): The topology configuration to run the operation in. Valid options are Ring and Linear. Default = Ring
```
Please align with @yan-zaretskiy om the new docs format. This op seem to be using the old format.

SeanNijjar replied:
Correct - the CCL ops have not been updated. I've opened an issue here to track: https://github.com/tenstorrent/tt-metal/issues/13249

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/host/reduce_scatter_common.hpp:
```diff
@@ -0,0 +1,84 @@
+
+struct WorkerTransferInfo {
+    WorkerTransferInfo(
+        std::vector<uint32_t> pages_per_full_chunk_per_worker, uint32_t num_links, uint32_t num_workers);
```
const ref for vector

SeanNijjar replied:
Thanks - this slipped through the cracks

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/host/reduce_scatter_common.cpp:
```diff
@@ -0,0 +1,173 @@
+
+    // Validate the worker attributes
+    for (const auto &wa : worker_attributes) {
+        TT_ASSERT(wa.send_to_edm_semaphore_id.has_value() || wa.receive_from_edm_semaphore_id.has_value(), "Internal error. Incorrectly setup worker attributes for reduce scatter");
```
Why is it an assert vs fatal here?

SeanNijjar replied:
Yea this is one of those things with assert vs fatal. Here's how I see it - let me know if you disagree.

If it's an external user facing API or some variable directly derived from some user provided information --> Always fatal if we know the provided input is incorrect/will provide issues with the program.

If it's an internal logic bug we are trying to catch --> Use TT_ASSERT as we should be catching these during testing... Unless these are disabled during our test builds too  . In this case I am trying to catch the internal logic bug category of error.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:
```diff
@@ -88,8 +91,8 @@ Tensor reduce_scatter(
             const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
-
-            bool is_ring = topology ==ttnn::ccl::Topology::Ring;
+            TT_ASSERT(input_tensors.size() >= 1, "Reduce scatter op expects an input tensor");
```
In most cases we should use Fatal (not a fatal really, throws an exception) instead of assert (only checks in debug build)

SeanNijjar replied:
updated

ayerofieiev-tt reviewed (approved):
Don’t have much to say for the ccl changes, but generally looks good. Left minor comments.

SeanNijjar committed:
#0: update license headers

SeanNijjar committed:
#0: disable debug mode for reduce scatter post commit

SeanNijjar committed:
#0: additional cleanup based on feedback:

- cleanup superfluous includes
- pass some things by ref
- assert -> fatal in some cases

SeanNijjar committed:
#0: sq

SeanNijjar committed:
#0: undo test disable

SeanNijjar committed:
#0: add missing license header

SeanNijjar committed:
#0: cleanup

SeanNijjar committed:
#11178: enable initial support for reduce scatter (limited functionality)

this commit adds the initial support for reduce scatter. However, only a
few cases are functional. Future work will improve correctness across
more cases.

======= Line Reduce Scatter Algorithm ======
The algorithm for line reduce scatter will send the minimal ammount of
data over each line and out of each chip. All diagrams are for an
example 4-chip line reduce scatter.

First the operation fractures each input tensor

Input Tensors ---------------
       |      |      |      |
       |      |      |      |
       v      v      v      v
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|

Chip   0      1      2      3

                 |
                 | Fracture
                 | Tensors
                 v

Input Tensors ---------------
       |      |      |      |
       |      |      |      |
       v      v      v      v
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|
      | |    | |    | |    | |
      |-|    |-|    |-|    |-|

Chip   0      1      2      3

With fracture tensors are reduced and collapsed to the diagonal across
the chips where the diagonal shows how the fractures spatially map to
the final outputs. For example, the first output is generated by
reducing the top chunk of each input tensor.

The reduction is performed by having each chip forward its input to its
neighbour. For chips that are not at the end of the line, they reduce
with their input and forward.

      |-|    |-|    |-|    |-|
      |#|<---| |<---| |<---| |
      |-|    |-|    |-|    |-|
      | |--->|#|<---| |<---| |
      |-|    |-|    |-|    |-|
      | |--->| |--->|#|<---| |
      |-|    |-|    |-|    |-|
      | |--->| |--->| |--->|#|
      |-|    |-|    |-|    |-|

Chip   0      1      2      3

However, note that each arrow from the diagram heading out of of a chip
in a given direction shares ethernet resources for all other arrows
heading in the same direction from that chip. This means there is
inherently serialization here. For that reason, we schedule the chunks
in some way.

The general scheduling strategy is to send the chunks that are furthest
from the final reduce output first and step through chunks that are
incrementally closer to the final output.

Each direction from a chip can be processed independently.

The diagram below is annotated with the "timesteps" when each chunk is
sent. Each timestep is marked relative to the chunk source.

      |-| t=0|-| t=0|-| t=0|-|
      |#|<---| |<---| |<---| |
      |-|t=2 |-| t=1|-| t=1|-|
      | |--->|#|<---| |<---| |
      |-|t=1 |-|t=1 |-| t=2|-|
      | |--->| |--->|#|<---| |
      |-|t=0 |-|t=0 |-|t=0 |-|
      | |--->| |--->| |--->|#|
      |-|    |-|    |-|    |-|

Chip   0      1      2      3

Finally, not that the final output requires a reduction from both
directions. Given that the two directions of the line are executing
completely indepdently, we require some sort of merge operation. At the
time of this commit, the merge strategy is to designate a master and
slave reducer direction. We arbitrarily choose the 'right' or
'clockwise' direction as the master.

The master direction will write its output to the output tensor but note
that this will only be a partial output. The slave direction will read
from the output tensor to merge with the data from producer chip. It
will read from the output tensor based on a credit passing from master
 (implemented via semaphores)

     -------Input Tensor
     |
     |
     |
     |
     |
     |      |---------|----------
     |----> |         |         |
            | Reader  | Sender  |---
  From EDM  | (master)| (master)|  |
 ---------> |         |         |  |
            |---------|---------|  |
                                   |
                                   |
        |------  Output Tensor <---|
        |                 ^
        |                 |---------
        |                          |
        |   |---------|----------  |
        --> |         |         |  |
            | Reader  | Sender  |--|
  From EDM  | (master)| (master)|
 ---------> |         |         |
            |---------|---------|

As a part of the line reduce scatter implementation, new CCL
componenst were added: ccl send and ccl command generators/readers.

The ccl_send (kernel) was used to implement the starting ends of the lines
(i.e. the first senders). Although the ccl_send provides more generic
send capabilities than line reduce scatter currently requires, it was
chosen because it is a basic building block also for future CCL
send/recv "operations" and higher level CCL programming models.

======= CCL Send (Kernel) =======
The ccl_send kernel acts like an interpreter of CCL commands. CCL
commands are, so far, limited to be a send from tensor to EDM of a
tensor slice. The command specifies some information about the tensor
(shape, slice/view shape, view offset, etc.).

CCL send is capable of executing multiple commands back to back. In the
context of line reduce scatter, the ccl_send implements the separate
sends of the fractured chunks on the left and right ends of the line. To
do this for a line reduce scatter, we invoke n commands where n=#chips
in the line. Future commands will let an invoker specify this basic
pattern as a single command.

Looking at the third diagram that outlines the timesteps for each chunk,
for the left/right tensors, each timestep directly maps to a separate
ccl command.

======= CCL Command Generators/Readers =======
To facilitate command generation, initial components have been added to
let the host serialize commands for the the ccl_send kernel.
Correspondingly, command unpacking logic is also specified for each
command. This is used to help simplify command generation for the host.

Note that ccl_send as a standalone kernel and operation is experimental
and has several limitations:
- Slice reads currently constrained to page aligned slices
- Host command generation doesn't support proper 4D shape support
    (although the kernel side will internally represent shapes as 4D)
- Only one command is currently supported (send tensor slice to EDM)
